.github/workflows/pr-gate.yml

```diff
@@ -0,0 +1,208 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# ---------------------------------------------------------
+name: pr-gate
+
+on:
+  push:
+    branches: [ laserprec/ghaction-ci ]
+
+# This file defines following CI workflow:
+#
+#  ┌──────────┐   ┌─────────┐   ┌────────┐
+#  │  static  ├─┬─►  build* ├─┬─►  test  │
+#  │ analysis │ │ │  (cpu)  │ │ │ report │
+#  └──────────┘ │ └─────────┘ │ └────────┘         
+#               │ ┌─────────┐ │
+#               ├─►  build* ├─┤
+#               │ │ (spark) │ │
+#               │ └─────────┘ │
+#               │ ┌─────────┐ │
+#               └─►  build* ├─┘
+#                 │  (gpu)  │   <-- TODO: Coming Soon
+#                 └─────────┘
+#                    ....
+# *each runs in PARALLEL different combinations
+#  of python version, OS, test subsets, etc
+#
+#  ASCII chart created via https://asciiflow.com/
+
+jobs:
+###############################################
+############### STATIC-ANALYSIS ###############
+###############################################
+  static-analysis:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Use Python 3.6
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.6
+    
+    - name: Install dependencies (tox)
+      run: |
+        python -m pip install --upgrade pip setuptools wheel
+        pip install tox
+
+    - name: Run flake8
+      # TODO: re-enable this flake8 blocks (turned off to get a draft of the pipeline infrastracture)
+      continue-on-error: true
+      run: |
+       tox -e flake8
+
+###############################################
+################## CPU-BUILD ##################
+###############################################
+  build-cpu:
+    runs-on: ${{ matrix.os }}
+    needs: static-analysis
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        python: [3.6]
+        # different kind of tests are located in tests/<unit|integration|smoke> folders
+        test-kind: ['unit'] 
+        # pytest markers configured in tox.ini. See https://docs.pytest.org/en/6.2.x/example/markers.html
+        test-marker: ['not gpu and not spark and not notebooks', 'not gpu and notebooks and not spark']
+
+    steps:
+    - uses: actions/checkout@v2
+    ################# Run Python tests #################
+    - name: Use Python ${{ matrix.python }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python }}
+    
+    - name: Install build dependencies (tox)
+      run: |
+        python -m pip install --upgrade pip setuptools wheel
+        pip install tox
+    
+    - name: Run ${{ matrix.test-kind }} tests ('${{ matrix.test-marker }}')
+      # '-e py' will use the default 'python' executable found in system path
+      # for why using tox, see: https://tox.readthedocs.io/en/latest/index.html
```

Comment:
my understanding is a new env will be set up for each tox task, given the number of dependencies we have this will add some overhead, do you know what the timing breakdown is for environment creation vs test execution for different matrix options?

Reply:
Right now all of the env matrix installs all of the extra dependencies (i.e. pip install recommenders['all']). I think the overhead is around 4min to install it. I am re-running the pipeline to confirm and give you the break down. I know we can map subsets of dependencies for the different test markers (e.g. recommender['spark'] -> "spark and notebooks"), but this will introduce new complexity to the pipeline and I want to avoid it.

Github action does offer [dependency cache](https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows), and I think it should work fine with tox. This can help alleviating this bottleneck.

Reply:
ah, the dependency cache is cool. ok sounds good.
i do think we will eventually want to test each of the separate dependency sets to make sure issues are not getting swept under the rug somehow during sub-dependency resolution when using 'all'. but for now i think it makes sense to start with the simple case.

Reply:
I agree too. I can update the feature roadmap to include this:

1. Propose and draft the CI pipeline in Github Action <---- (Current PR)
2. Optimize build time for unit tests
3. Enforce flake8 (coding style checks) on the build and clean up coding styles to pass flake8 checks
4. Setup self-hosted machines to run our GPU workloads
5. **Run tests on the appropriate dependency subsets** 
6. Create feature parity with the existing CI pipelines (pr-gates & night-build)
7. Deprecate CI pipelines in ADO and switch to Github Actions

Reply:
Installing all the dependencies takes around ~5min in the tox build on the agent.
![image](https://user-images.githubusercontent.com/19534002/129041886-9ed51b53-ccf9-456a-8883-7d466321400b.png)
![image](https://user-images.githubusercontent.com/19534002/129043327-a61d176a-d2c3-4730-92d3-aeca9c82102f.png)